In [1]:
from pyspark.sql.functions import col, split, concat_ws, when, size, min, trim, round, mean, lit, max, abs

In [2]:
from pyspark.ml.feature import VectorAssembler
from pyspark.ml.regression import GBTRegressor
from pyspark.ml.evaluation import RegressionEvaluator

In [3]:
import findspark
findspark.init()

from pyspark.sql import SparkSession

spark = SparkSession.builder \
    .master("local[*]") \
    .appName("TrackTimePrediction") \
    .getOrCreate()

spark

Using Spark's default log4j profile: org/apache/spark/log4j2-defaults.properties
25/06/20 01:04:18 WARN Utils: Your hostname, Elijahs-MacBook-Pro.local, resolves to a loopback address: 127.0.0.1; using 192.168.1.184 instead (on interface en0)
25/06/20 01:04:18 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address
Using Spark's default log4j profile: org/apache/spark/log4j2-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
25/06/20 01:04:22 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable


In [27]:
spark.stop()

Machine learning time

In [4]:
file_path1 = "./final_database/better_merge_five.csv"
df = spark.read.csv(file_path1, header=True, inferSchema=True)
df = df.filter(col("PR_seconds") >= 700)
print(df.count())
df.show(1)

2893
+---+----------+--------------------+------------+-----+------+-----+-----+-----------------+-----+-------+-------+-------+-------+-------+-------+-------+-------+------+------+------+------+---------------+------------+-------+----------------+----------+-------------+
|_c0|Unnamed: 0|      Athlete/School|hs_grad_year|State|Gender|FR_5k|SO_5k|            JR_5k|SR_5k|FR_3200|SO_3200|JR_3200|SR_3200|FR_1600|SO_1600|JR_1600|SR_1600|FR_800|SO_800|JR_800|SR_800|hs_athlete_name|athlete_name|     id|      PR_seconds|hs_grad_yr|hs_grad_yr_gt|
+---+----------+--------------------+------------+-----+------+-----+-----+-----------------+-----+-------+-------+-------+-------+-------+-------+-------+-------+------+------+------+------+---------------+------------+-------+----------------+----------+-------------+
|354|       354|Adrian Diaz Bisho...|        2016|   CA|  Boys| NULL| NULL|935.6833333333334|920.7|   NULL|  588.5|  562.1|  559.0|   NULL|   NULL|  263.9|  265.5|  NULL|  NULL|  NUL

25/06/20 01:04:30 WARN SparkStringUtils: Truncated the string representation of a plan since it was too large. This behavior can be adjusted by setting 'spark.sql.debug.maxToStringFields'.


To deal with NULLs, will try imputing with an indicator column, and simply setting the NULLs to a very large value. The first is more robust, but I worry about setting nulls to the mean since I am hoping to capture trends

In [5]:
df = df.fillna(9999999, subset=['FR_5k', 'SO_5k', 'JR_5k', 'SR_5k', 'FR_3200', 'SO_3200', 'JR_3200', 'SR_3200', 'FR_1600', 'SO_1600', 'JR_1600', 'SR_1600', 'FR_800', 'SO_800', 'JR_800', 'SR_800'])
df.show(5)

+-----+----------+--------------------+------------+-----+------+---------+---------+-----------------+---------+---------+-------+---------+-------+---------+---------+---------+-------+---------+---------+---------+---------+---------------+-------------+-------+------------------+----------+-------------+
|  _c0|Unnamed: 0|      Athlete/School|hs_grad_year|State|Gender|    FR_5k|    SO_5k|            JR_5k|    SR_5k|  FR_3200|SO_3200|  JR_3200|SR_3200|  FR_1600|  SO_1600|  JR_1600|SR_1600|   FR_800|   SO_800|   JR_800|   SR_800|hs_athlete_name| athlete_name|     id|        PR_seconds|hs_grad_yr|hs_grad_yr_gt|
+-----+----------+--------------------+------------+-----+------+---------+---------+-----------------+---------+---------+-------+---------+-------+---------+---------+---------+-------+---------+---------+---------+---------+---------------+-------------+-------+------------------+----------+-------------+
|  354|       354|Adrian Diaz Bisho...|        2016|   CA|  Boys|99999

In [6]:
feature_cols = ['FR_5k', 'SO_5k', 'JR_5k', 'SR_5k', 'FR_3200', 'SO_3200', 'JR_3200', 'SR_3200', 'FR_1600', 'SO_1600', 'JR_1600', 'SR_1600', 'FR_800', 'SO_800', 'JR_800', 'SR_800']
assembler = VectorAssembler(inputCols=feature_cols, outputCol="features", handleInvalid='keep')
model_df = assembler.transform(df)
model_df.select('features', 'PR_seconds').show(5, truncate=False)

+-------------------------------------------------------------------------------------------------------------------------------------------------+------------------+
|features                                                                                                                                         |PR_seconds        |
+-------------------------------------------------------------------------------------------------------------------------------------------------+------------------+
|[9999999.0,9999999.0,935.6833333333334,920.7,9999999.0,588.5,562.1,559.0,9999999.0,9999999.0,263.9,265.5,9999999.0,9999999.0,9999999.0,9999999.0]|902.539999961853  |
|[9999999.0,9999999.0,1093.4,1037.4,9999999.0,683.8,9999999.0,623.9,9999999.0,303.4,294.6,287.2,9999999.0,138.5,9999999.0,126.5]                  |932.5400009155273 |
|[984.3,995.1,978.2,941.3,585.8,594.9,557.2,573.5,269.6,268.8,257.3,268.7,127.1,9999999.0,118.4,9999999.0]                                        |872.3899993896484 

In [7]:
train_data, test_data = model_df.randomSplit(weights=[0.8, 0.2], seed=27)
print(f"Training Dataset Count: {train_data.count()}")
print(f"Test Dataset Count: {test_data.count()}")

Training Dataset Count: 2294
Test Dataset Count: 599


In [8]:
gbt = GBTRegressor(featuresCol='features', labelCol='PR_seconds')
print("Training the GBT model...")
gbt_model = gbt.fit(train_data)
print("Training complete.")

Training the GBT model...
Training complete.


In [9]:
predictions = gbt_model.transform(test_data)
predictions.select("PR_seconds", "prediction").show(10)

+-----------------+-----------------+
|       PR_seconds|       prediction|
+-----------------+-----------------+
|919.0499992370605|891.7959749679327|
|873.5900001525879|894.6008986106523|
|853.1199998855591|916.1585059152549|
|865.6399993896484|919.5875773792714|
|854.3000001907349|854.1644066830344|
|874.4900016784668|881.8998099477036|
| 964.039999961853|939.4477882701136|
|969.5299997329712|946.8188193520594|
|898.5400009155273|910.1880148899103|
|956.6300010681152|921.7046155768478|
+-----------------+-----------------+
only showing top 10 rows


25/06/20 01:05:02 WARN InstanceBuilder: Failed to load implementation from:dev.ludovic.netlib.blas.JNIBLAS


In [10]:
evaluator = RegressionEvaluator(
    labelCol="PR_seconds",
    predictionCol="prediction",
    metricName="rmse"  
)
rmse = evaluator.evaluate(predictions)
print(f"Root Mean Squared Error (RMSE) on test data = {rmse:.2f} seconds")
r2_evaluator = evaluator.setMetricName("r2")
r2 = r2_evaluator.evaluate(predictions)
print(f"R-squared (R2) on test data = {r2:.2f}")


Root Mean Squared Error (RMSE) on test data = 43.36 seconds
R-squared (R2) on test data = 0.36


This model sucks. It is on average 62.94 seconds off. Will try to impute with indicators next. Then add more features like progression, but realistically, I just don't have enough data yet to make a broad enough model to account for all the factors at play. Will also try using the dataset with a lower threshold. (After removing the 100s 5k times, RMSE=49.50 seconds)

Using 5 thresh, RMSE = 46.91

better merging down to 43.36 RMSE

In [53]:
max_val = df.select(max(col("SR_800"))).first()[0]
df = df.withColumn(
    "SR_800_was_missing",
    when(col("SR_800").isNull(), 1).otherwise(0)
)
df = df.fillna(max_val, subset=["SR_800"])
df.show(1)

+--------------+---+----------+--------------------+-----+-----+------+------+------+-----------------+-----------------+-------+-------+-------+-------+-------+-------+-------+-------+------+------+------+------+-------+-----------------+-----------------+-----------------+-----------------+-----------------+-------------------+-------------------+-------------------+-------------------+-------------------+-------------------+-------------------+-------------------+------------------+------------------+------------------+------------------+
|  athlete_name|_c0|Unnamed: 0|      Athlete/School|Grade|State|Gender| FR_5k| SO_5k|            JR_5k|            SR_5k|FR_3200|SO_3200|JR_3200|SR_3200|FR_1600|SO_1600|JR_1600|SR_1600|FR_800|SO_800|JR_800|SR_800|     id|       PR_seconds|FR_5k_was_missing|SO_5k_was_missing|JR_5k_was_missing|SR_5k_was_missing|FR_3200_was_missing|SO_3200_was_missing|JR_3200_was_missing|SR_3200_was_missing|FR_1600_was_missing|SO_1600_was_missing|JR_1600_was_missing|S

In [54]:
output_path = './final_database'
df.coalesce(1).write \
    .option('header', 'true') \
    .mode('append') \
    .csv(output_path) 

In [55]:
feature_cols = ['FR_5k', 'SO_5k', 'JR_5k', 'SR_5k', 'FR_3200', 'SO_3200', 'JR_3200', 'SR_3200', 'FR_1600', 'SO_1600', 'JR_1600', 'SR_1600', 'FR_800', 'SO_800', 'JR_800', 'SR_800', 'FR_5k_was_missing', 'SO_5k_was_missing', 'JR_5k_was_missing', 'SR_5k_was_missing', 'FR_3200_was_missing', 'SO_3200_was_missing', 'JR_3200_was_missing', 'SR_3200_was_missing', 'FR_1600_was_missing', 'SO_1600_was_missing', 'JR_1600_was_missing', 'SR_1600_was_missing', 'FR_800_was_missing', 'SO_800_was_missing', 'JR_800_was_missing', 'SR_800_was_missing']
assembler = VectorAssembler(inputCols=feature_cols, outputCol="features", handleInvalid='keep')
df = df.filter(col("PR_seconds") >= 700)
model_df = assembler.transform(df)
model_df.select('features', 'PR_seconds').show(5, truncate=False)

+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------+
|features                                                                                                                                                                                   |PR_seconds        |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------+
|[1212.0,1250.4,961.5166666666668,952.7333333333336,782.5,600.4,727.0,559.4,337.2,323.9,269.0,260.7,150.3,140.4,142.0,123.9,1.0,1.0,0.0,0.0,1.0,0.0,1.0,0.0,1.0,1.0,0.0,0.0,1.0,1.0,1.0,0.0]|919.0499992370605 |
|[1212.0,985.8,926.9,915.0166666666668,782.5,597.0,574.8,567.8,337.2,323.9,267.6,264.6,150.3,140.4,142.0,152.3,1.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,1.0,1.0,0.0,0.0,1.0,1

In [56]:
train_data, test_data = model_df.randomSplit(weights=[0.8, 0.2], seed=27)
print(f"Training Dataset Count: {train_data.count()}")
print(f"Test Dataset Count: {test_data.count()}")

Training Dataset Count: 2914
Test Dataset Count: 761


In [57]:
gbt = GBTRegressor(featuresCol='features', labelCol='PR_seconds')
print("Training the GBT model...")
gbt_model = gbt.fit(train_data)
print("Training complete.")

Training the GBT model...
Training complete.


In [58]:
predictions = gbt_model.transform(test_data)
predictions.select("PR_seconds", "prediction").show(20)

+------------------+-----------------+
|        PR_seconds|       prediction|
+------------------+-----------------+
| 917.2800006866455|920.4076651790483|
| 919.0499992370605|904.8468310476235|
| 909.7799997329712|917.6375076357772|
| 940.8600006103516|924.8858759982098|
| 953.8800010681152|928.8988816005892|
| 955.7400016784668|977.2593266884074|
| 876.2299995422363|876.5921103892258|
| 859.1499996185303|877.6421014457443|
| 898.8899993896484|888.4804398747314|
| 893.9300003051758|946.0214884550903|
| 944.9199981689453|942.0453454809947|
| 920.9599990844727| 919.192528016944|
| 930.4699993133545|981.7547963022289|
|  893.439998626709|  921.99904595892|
|1072.4599990844727|978.2544579064887|
| 879.2000007629395|956.3157814492938|
| 875.2299995422363|898.5304228532132|
| 885.7400016784668|927.8638224552273|
| 969.8500003814697|915.5752108894388|
| 893.6199989318848|902.7313686508827|
+------------------+-----------------+
only showing top 20 rows


In [59]:
evaluator = RegressionEvaluator(
    labelCol="PR_seconds",
    predictionCol="prediction",
    metricName="rmse"  
)
rmse = evaluator.evaluate(predictions)
print(f"Root Mean Squared Error (RMSE) on test data = {rmse:.2f} seconds")
r2_evaluator = evaluator.setMetricName("r2")
r2 = r2_evaluator.evaluate(predictions)
print(f"R-squared (R2) on test data = {r2:.2f}")

Root Mean Squared Error (RMSE) on test data = 46.97 seconds
R-squared (R2) on test data = 0.26


In [16]:
predictions_with_error = predictions.withColumn(
    "error_seconds",
    abs(col("PR_seconds") - col("prediction"))
)
predictions_with_error.select("PR_seconds", "prediction", "error_seconds").show(20)

+------------------+-----------------+------------------+
|        PR_seconds|       prediction|     error_seconds|
+------------------+-----------------+------------------+
| 919.0499992370605|891.7959749679327|27.254024269127854|
| 873.5900001525879|894.6008986106523| 21.01089845806439|
| 853.1199998855591|916.1585059152549| 63.03850602969578|
| 865.6399993896484|919.5875773792714|53.947577989622914|
| 854.3000001907349|854.1644066830344|0.1355935077004915|
| 874.4900016784668|881.8998099477036| 7.409808269236805|
|  964.039999961853|939.4477882701136|24.592211691739408|
| 969.5299997329712|946.8188193520594|22.711180380911742|
| 898.5400009155273|910.1880148899103|11.648013974382934|
| 956.6300010681152|921.7046155768478|34.925385491267434|
|1004.3800010681152|921.7659771430112|   82.614023925104|
| 904.7600002288818| 920.598150596273|15.838150367391108|
| 861.4899997711182|839.2617715382629| 22.22822823285526|
| 846.4499998092651|917.0033758889754| 70.55337607971023|
| 893.00999832

In [19]:
error_threshold = 60

# Filter for rows where the error is greater than your threshold
outliers_df = predictions_with_error.filter(col("error_seconds") > error_threshold) \
    .orderBy(col("error_seconds").desc()) # Sort to see the biggest errors first

print(f"Found {outliers_df.count()} outlier predictions (error > {error_threshold} seconds).")

# Display the top outliers
print("\nRows with the largest prediction errors:")
outliers_df.show(20, truncate=False)

Found 85 outlier predictions (error > 60 seconds).

Rows with the largest prediction errors:
+-----+----------+--------------------------------------+------------+-----+------+---------+------------------+-----------------+-----------------+---------+---------+---------+---------+---------+---------+---------+---------+---------+---------+---------+---------+-----------------+-----------------+-------+------------------+----------+-------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------+------------------+
|_c0  |Unnamed: 0|Athlete/School                        |hs_grad_year|State|Gender|FR_5k    |SO_5k             |JR_5k            |SR_5k            |FR_3200  |SO_3200  |JR_3200  |SR_3200  |FR_1600  |SO_1600  |JR_1600  |SR_1600  |FR_800   |SO_800   |JR_800   |SR_800   |hs_athlete_name  |athlete_name     |id     |PR_seconds        |hs_grad_yr|hs

25/06/16 21:01:04 WARN HeartbeatReceiver: Removing executor driver with no recent heartbeats: 679030 ms exceeds timeout 120000 ms
25/06/16 21:01:04 WARN SparkContext: Killing executors is not supported by current scheduler.
25/06/16 21:01:11 ERROR Inbox: Ignoring error
org.apache.spark.SparkException: Exception thrown in awaitResult: 
	at org.apache.spark.util.SparkThreadUtils$.awaitResult(SparkThreadUtils.scala:53)
	at org.apache.spark.util.ThreadUtils$.awaitResult(ThreadUtils.scala:342)
	at org.apache.spark.rpc.RpcTimeout.awaitResult(RpcTimeout.scala:75)
	at org.apache.spark.rpc.RpcEnv.setupEndpointRefByURI(RpcEnv.scala:102)
	at org.apache.spark.rpc.RpcEnv.setupEndpointRef(RpcEnv.scala:110)
	at org.apache.spark.util.RpcUtils$.makeDriverRef(RpcUtils.scala:36)
	at org.apache.spark.storage.BlockManagerMasterEndpoint.driverEndpoint$lzycompute(BlockManagerMasterEndpoint.scala:132)
	at org.apache.spark.storage.BlockManagerMasterEndpoint.org$apache$spark$storage$BlockManagerMasterEndpoint$$

In [ ]:
saved_model_path = './gbt_model'
gbt_model.write().overwrite().save(saved_model_path)

25/06/20 07:45:46 WARN HeartbeatReceiver: Removing executor driver with no recent heartbeats: 1018940 ms exceeds timeout 120000 ms
25/06/20 07:45:46 WARN SparkContext: Killing executors is not supported by current scheduler.
25/06/20 07:45:53 ERROR Inbox: Ignoring error
org.apache.spark.SparkException: Exception thrown in awaitResult: 
	at org.apache.spark.util.SparkThreadUtils$.awaitResult(SparkThreadUtils.scala:53)
	at org.apache.spark.util.ThreadUtils$.awaitResult(ThreadUtils.scala:342)
	at org.apache.spark.rpc.RpcTimeout.awaitResult(RpcTimeout.scala:75)
	at org.apache.spark.rpc.RpcEnv.setupEndpointRefByURI(RpcEnv.scala:102)
	at org.apache.spark.rpc.RpcEnv.setupEndpointRef(RpcEnv.scala:110)
	at org.apache.spark.util.RpcUtils$.makeDriverRef(RpcUtils.scala:36)
	at org.apache.spark.storage.BlockManagerMasterEndpoint.driverEndpoint$lzycompute(BlockManagerMasterEndpoint.scala:132)
	at org.apache.spark.storage.BlockManagerMasterEndpoint.org$apache$spark$storage$BlockManagerMasterEndpoint$

Imputing with the mean and marking which rows were missing was worse: 74.15 seconds off on average. Will test the other data set next and add new features soon. (I accidentally had some 800m results as 5k times, new RMSE = 49.93). 49.28 using MAX instead of mean.

five thresh RMSE 46.97